File: src_fortran/sub/backup.f
c
c
c
c
c
c
      subroutine BACKUP
c
c     go back in direction of the previous solution by doing
c     only half of the adjustments calculated in the last iteration
c
      implicit none
      include '../inc/vel_com.inc'
      integer jjj,i,n,k,iccc,k2,nl,j1,j2,ifl,k1,ksta1,j,kk1,m
      integer ksta2
      real zzz,avelo
c
      real cc(ist)
      character*1 reflch
c
      jjj=0
      do 2 i=1,legs
c---------- hypocenter backup
      n=4
c
      if(i.gt.neqs) n=1
      do 2 k=1,n
      jjj=jjj+1
      b(jjj)=b(jjj)/2.0
      if(k.eq.4)then                      ! backup depth
c
c     concept of effdeltaz no longer in use!
c
c         if(effdeltaz(i).eq.0.0)then
            e(k,i)=e(k,i)-b(jjj)          ! depth not constrained
c         else
c            effdeltaz(i)=effdeltaz(i)/2.0
c            e(k,i)=e(k,i)-effdeltaz(i)           !  depth was constrained:
c         endif                                   ! go back only half of the
         if(itopo.gt.0.and.e(k,i).lt.0.0)then
c           depth above zero... is it also above surface?
            call CHTOP(-e(2,i),e(3,i),zzz,
     &                 topo1file,topo2file)
            if(e(k,i).lt.zzz) e(k,i)=zzz  ! set hypocenter down to surface !
            if(ifixsolution.gt.0) e(k,i)=zzz ! fix depth to min_depth allowed!!
         endif
      else                                       ! effective depth-change !
            e(k,i)=e(k,i)-b(jjj)
      endif
   2  continue
ccc      iccc=mod((nitt-1),invertratio)
      iccc=mod(nitt,invertratio)
      if(iccc.ne.0) goto 510
c---- velocity readjustments for each model
      do 26 k2=1,nmod
      nl=nplay(k2)
      j1=4*neqs+nshot+laysum(k2)
      j2=j1-1+nl
      k=0
      if(.not.single_turbo)then
         write(16,3)
3        format(1x,'Velocity readjustments:')
         write(16,27) k2
27       format(1x,'Velocity model',i4)
      endif
      do 4 jjj=j1,j2
      k=k+1
      b(jjj)=b(jjj)/2.
      vp(k2,k)=vp(k2,k)-b(jjj)
      if(lowveloclay.eq.0)then
         if(k.gt.1)then
            if(vp(k2,k).lt.vp(k2,k-1))then
               i f (.not.single_turbo)then
               write(16,*)'WARNING:'
               write(16,'(1x,''WARNING: Tried to introduce a '',
     &                    ''low-velocity-layer! (Layer '',i2,'')'')') k
               write(16,'(1x,''Setting DVP from '',f5.2,'' to 0.0 and'',
     &                    '' VP to vp(layer_above)+0.001'')') b(jjj)
               e n d i f
               write(6,'(1x,''WARNING: Tried to introduce a '',
     &                    ''low-velocity-layer! (Layer '',i2,'')'')') k
               write(6,'(1x,''Setting DVP from '',f5.2,'' to 0.0 and'',
     &                    '' VP to vp(layer_above)+0.001'')') b(jjj)
               write(6,*)
               b(jjj)=0.0
               vp(k2,k)=vp(k2,k-1)+0.001
            endif
         endif
      endif
5     format(1x,3f7.3,3x,a1)
      if(k.eq.ireflector)reflch=reflchar
      if(k.ne.ireflector)reflch=' '
      if(.not.single_turbo)then
         write(16,5) vp(k2,k),b(jjj),hp(k2,k),reflch
      endif
   4  continue
c
c    calculate and print average velocities of the model k2 :
c
      i f (.not.single_turbo) t h e n
      ifl=1
      write(16,*)
      write(16,*)'Calculation of average velocity starts at layer # ',
     &           ifl
      avelo=0
      do k=ifl+1,nplay(k2)
         avelo=avelo + ( hp(k2,k)-hp(k2,k-1) ) * vp(k2,k-1)
         write(16,'('' z ='',f6.2,'' ...'',f6.2,''    v = '',f4.2,
     &             ''    AVEL = '',f4.2,'' downto z = '',f6.2)')
     &       hp(k2,k-1),hp(k2,k),vp(k2,k-1),avelo/(hp(k2,k)-hp(k2,ifl)),
     &       hp(k2,k)
      enddo
      write(16,*)
c
      ifl=2
      write(16,*)
      write(16,*)'Calculation of average velocity starts at layer # ',
     &           ifl
      avelo=0
      do k=ifl+1,nplay(k2)
         avelo=avelo + ( hp(k2,k)-hp(k2,k-1) ) * vp(k2,k-1)
         write(16,'('' z ='',f6.2,'' ...'',f6.2,''    v = '',f4.2,
     &             ''    AVEL = '',f4.2,'' downto z = '',f6.2)')
     &       hp(k2,k-1),hp(k2,k),vp(k2,k-1),avelo/(hp(k2,k)-hp(k2,ifl)),
     &       hp(k2,k)
      enddo
      write(16,*)
      write(16,*)
      e n d i f
c
26    continue
510   if(nsinv.eq.0) goto 511
ccc      if(mod((nitt-1),invertratio).ne.0) goto 511
      if(mod(nitt,invertratio).ne.0) goto 511
      k1=4*neqs+nshot+nltot+1
      ksta1=ksta
      if(nsp.eq.2) ksta1=(ksta/2)
      k2=k1+ksta1-1
      do 60 j=k1,k2
60    b(j)=b(j)/2.
      do 7 j=1,nsta
      cc(j)=0.0
      if(map1(j).eq.0) goto 7
      if(map1(j).gt.ksta1) goto 7
      kk1=k1-1+map1(j)
      cc(j)=b(kk1)
      ptcor(j)=ptcor(j)-cc(j)
7     continue
      if(.not.single_turbo)then
         write(16,8) (stn(m),ptcor(m),cc(m),m=1,nsta)
8        format(5(2x,a4,2f7.3))
 6       format(1x,'P correction readjustments:')
         write(16,*)
         write(16,*)'Half adjustments made'
      endif
      if(nsp.ne.2) goto 29
      k1=4*neqs+nshot+nltot+ksta1+1
      ksta2=ksta-ksta1
      k2=k1+ksta2-1
      do 61 j=k1,k2
61    b(j)=b(j)/2.
      if(.not.single_turbo)then
         write(16,39)
39       format(1x,'S correction readjustments:')
      endif
      do 28 j=1,nsta
      cc(j)=0.0
      if(map1(j).eq.0) goto 28
      if(map1(j).gt.ksta2) goto 28
      kk1=k1-1+map1(j)
      cc(j)=b(kk1)
      stcor(j)=stcor(j)-cc(j)
28    continue
      if(.not.single_turbo)then
         write(16,8) (stn(m),stcor(m),cc(m),m=1,nsta)
      endif
511   continue
      if(.not.single_turbo)then
         write(16,*)
         write(16,*)'Half adjustments made'
      endif
29    continue
c
      return
      end ! of subr. backup

File: src/sub/backup.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#define IEQ 658

typedef void (*chtop_backup_fn_t)(double x, double y, double *zmin,
                                  const char *topo1file, const char *topo2file);

void backup_model(
    int legs,
    int neqs,
    int nshot,
    int nitt,
    int invertratio,
    int nmod,
    int nltot,
    int nsta,
    int nsp,
    int nsinv,
    int ksta,
    int itopo,
    int ifixsolution,
    int lowveloclay,
    int ireflector,
    char reflchar,
    const int *nplay,
    const int *laysum,
    const int *map1,
    double *b,
    double e[][IEQ],
    double *vp,
    const double *hp,
    double *ptcor,
    double *stcor,
    bool single_turbo,
    FILE *logfp,
    FILE *warnfp,
    const char (*stn)[5],
    const char *topo1file,
    const char *topo2file,
    chtop_backup_fn_t chtop_fn) {
    int i;
    int j;
    int k;
    int jjj = -1;
    double *cc;

    if (legs <= 0 || neqs < 0 || nsta < 0 || nmod < 0 || nltot < 0 ||
        invertratio == 0 || nplay == NULL || laysum == NULL || map1 == NULL ||
        b == NULL || e == NULL || vp == NULL || hp == NULL || ptcor == NULL ||
        stcor == NULL) {
        return;
    }

    if (!single_turbo && logfp == NULL) {
        logfp = stdout;
    }
    if (warnfp == NULL) {
        warnfp = stderr;
    }

    cc = (double *)calloc((size_t)nsta, sizeof(double));
    if (cc == NULL) {
        return;
    }

    for (i = 0; i < legs; ++i) {
        int n = (i < neqs) ? 4 : 1;
        for (k = 0; k < n; ++k) {
            ++jjj;
            b[jjj] /= 2.0f;

            if (k == 3) {
                double zzz = 0.0f;
                e[k][i] -= b[jjj];

                if (itopo > 0 && e[k][i] < 0.0f && chtop_fn != NULL) {
                    chtop_fn(-e[1][i], e[2][i], &zzz, topo1file,
                             topo2file);
                    if (e[k][i] < zzz) {
                        e[k][i] = zzz;
                    }
                    if (ifixsolution > 0) {
                        e[k][i] = zzz;
                    }
                }
            } else {
                e[k][i] -= b[jjj];
            }
        }
    }

    if ((nitt % invertratio) == 0) {
        for (i = 0; i < nmod; ++i) {
            int nl = nplay[i];
            int j1 = 4 * neqs + nshot + laysum[i];

            if (!single_turbo && logfp != NULL) {
                fprintf(logfp, " Velocity readjustments:\n");
                fprintf(logfp, " Velocity model%4d\n", i + 1);
            }

            for (k = 0; k < nl; ++k) {
                int bidx = j1 + k;
                char reflch = (k == (ireflector - 1)) ? reflchar : ' ';

                b[bidx] /= 2.0f;
                vp[i * nltot + k] -= b[bidx];

                if (lowveloclay == 0 && k > 0 &&
                    vp[i * nltot + k] < vp[i * nltot + (k - 1)]) {
                    if (!single_turbo && logfp != NULL) {
                        fprintf(logfp,
                                " WARNING: Tried to introduce a low-velocity-layer! (Layer %2d)\n",
                                k + 1);
                        fprintf(logfp,
                                " Setting DVP from %5.2f to 0.0 and VP to vp(layer_above)+0.001\n",
                                b[bidx]);
                    }
                    fprintf(warnfp,
                            " WARNING: Tried to introduce a low-velocity-layer! (Layer %2d)\n",
                            k + 1);
                    fprintf(warnfp,
                            " Setting DVP from %5.2f to 0.0 and VP to vp(layer_above)+0.001\n",
                            b[bidx]);
                    b[bidx] = 0.0f;
                    vp[i * nltot + k] = vp[i * nltot + (k - 1)] + 0.001f;
                }

                if (!single_turbo && logfp != NULL) {
                    fprintf(logfp, " %7.3f %7.3f %7.3f   %c\n", vp[i * nltot + k],
                            b[bidx], hp[i * nltot + k], reflch);
                }
            }
        }
    }

    if (nsinv != 0 && (nitt % invertratio) == 0) {
        int k1 = 4 * neqs + nshot + nltot;
        int ksta1 = (nsp == 2) ? (ksta / 2) : ksta;
        int k2 = k1 + ksta1;

        for (j = k1; j < k2; ++j) {
            b[j] /= 2.0f;
        }

        for (j = 0; j < nsta; ++j) {
            int m = map1[j];
            cc[j] = 0.0f;
            if (m < 0 || m >= ksta1) {
                continue;
            }
            cc[j] = b[k1 + m];
            ptcor[j] -= cc[j];
        }

        if (!single_turbo && logfp != NULL) {
            for (j = 0; j < nsta; ++j) {
                fprintf(logfp, "  %.4s %7.3f %7.3f", stn != NULL ? stn[j] : "    ",
                        ptcor[j], cc[j]);
                if ((j + 1) % 5 == 0 || j + 1 == nsta) {
                    fprintf(logfp, "\n");
                }
            }
            fprintf(logfp, "\n");
            fprintf(logfp, " Half adjustments made\n");
        }

        if (nsp == 2) {
            int k1s = 4 * neqs + nshot + nltot + ksta1;
            int ksta2 = ksta - ksta1;
            int k2s = k1s + ksta2;

            for (j = k1s; j < k2s; ++j) {
                b[j] /= 2.0f;
            }

            if (!single_turbo && logfp != NULL) {
                fprintf(logfp, " S correction readjustments:\n");
            }

            for (j = 0; j < nsta; ++j) {
                int m = map1[j];
                cc[j] = 0.0f;
                if (m < 0 || m >= ksta2) {
                    continue;
                }
                cc[j] = b[k1s + m];
                stcor[j] -= cc[j];
            }

            if (!single_turbo && logfp != NULL) {
                for (j = 0; j < nsta; ++j) {
                    fprintf(logfp, "  %.4s %7.3f %7.3f", stn != NULL ? stn[j] : "    ",
                            stcor[j], cc[j]);
                    if ((j + 1) % 5 == 0 || j + 1 == nsta) {
                        fprintf(logfp, "\n");
                    }
                }
            }
        }
    }

    if (!single_turbo && logfp != NULL) {
        fprintf(logfp, "\n");
        fprintf(logfp, " Half adjustments made\n");
    }

    free(cc);
}

extern int legs, neqs, nshot, nitt, invertratio, nmod, nltot, nsta, nsp, nsinv, ksta;
extern int itopo, ifixsolution, lowveloclay, ireflector;
extern char reflchar;
extern int nplay[];
extern int laysum[];
extern int map1[];
extern double b[];
extern double e[][658];
extern double vp[][100];
extern double hp[][100];
extern double ptcor[];
extern double stcor[];
extern bool single_turbo;
extern char stn[][5];
extern char topo1file[81], topo2file[81];
extern void chtop(double x, double y, double *zmin, const char *topo1file, const char *topo2file);

void backup(void) {
    backup_model(
        legs,
        neqs,
        nshot,
        nitt,
        invertratio,
        nmod,
        nltot,
        nsta,
        nsp,
        nsinv,
        ksta,
        itopo,
        ifixsolution,
        lowveloclay,
        ireflector,
        reflchar,
        nplay,
        laysum,
        map1,
        b,
        e,
        &vp[0][0],
        &hp[0][0],
        ptcor,
        stcor,
        single_turbo != 0,
        stdout,
        stderr,
        (const char (*)[5])stn,
        topo1file,
        topo2file,
        chtop);
}
